WikiWatershed/model-my-watershed
# Split NLCD and Soil Analyze Endpoints, Refactor Code

## Overview

Previously there was a single endpoint for getting land and soil surveys for an area of interest, to be used in Analyze view of the front-end. This reused the geoprocessing design for TR-55 for convenience. Unfortunately, this was a costly decision, for TR-55 geoprocessing service overlaid land and soil values and returned a combined result. This result was useful for TR-55, but not for Analyze, which only needs the discrete counts for each.

By switching to independent endpoints, we make Analyze much faster and simpler. As a consequence, the values can now no longer be shared with TR-55, causing it to start a new geoprocessing chain when it is triggered. However, this is an acceptable compromise, as the user may choose another model such as MapShed to run after Analyze, in which case the old values would be discarded anyway.

In addition to the split, a number of refactors have also been performed. The standardization to one geoprocessing back-end invites folding and reuse. The main refactoring here revolves around two ideas:

 * Since both MapShed and TR-55 use the same geoprocessing conventions, the relevant methods have been moved to the `geoprocessing` module
 * Since each `/analyze` endpoint now returns a single survey value, arrays have been replaced with direct access in both front- and back-ends

See commit messages for details.

Also tagging @mmcfarland and @kdeloach for review.

Connects #1877 

## Testing Instructions

 * Pick a well known AoI and run it on `develop` or [staging](https://staging.app.wikiwatershed.org/)
 * Pick the same AoI and run it on this branch
 * Ensure that Analyze results and TR-55 results are identical
 * Observe that Analyze is slightly faster than before, and TR-55 is slightly slower than before
 * Go through the entire process of selecting an area, analyzing it, modeling with TR-55, adding modifications, adjusting precipitation, and comparing models, and ensure that everything still works.
